// ts-glitter/src/config.ts
'use strict';
import path from "path";
import {get} from "underscore";
import dotenv from 'dotenv';
export class ConfigInitial{
    public static  envPath=path.resolve(__dirname, `../environments/${process.env.NODE_ENV || 'local'}.env`)
}
dotenv.config({
    path: path.resolve(__dirname, `../environments/${process.env.NODE_ENV || 'local'}.env`),
});
export const saasConfig = {
    SAAS_NAME: process.env.GLITTER_DB,
    DEF_DEADLINE: 7
}

export const config = {
    SECRET_KEY: 'dsklkmsadl',
    API_PREFIX: "/api/v1",
    PARAMS_NEED_ENCRYPT_IN_LOG: ['pwd', 'email'],
    PWD_SALT_ROUND: 5,
    SERVER_PORT: '3080',
    LOG_PATH: path.resolve("../../log"),
    /*Database*/
    DB_CONN_LIMIT: 1,
    DB_QUEUE_LIMIT: 0,
    DB_URL: process.env.DB_URL,
    DB_PORT: 3306,
    DB_USER: process.env.DB_USER,
    DB_PWD: process.env.DB_PWD,
    DB_NAME: process.env.GLITTER_DB,
    /*REDIS*/
    REDIS_URL: process.env.REDIS_URL,
    REDIS_PORT: process.env.REDIS_PORT,
    REDIS_PWD: process.env.REDIS_PWD,
    /*AWS*/
    AWS_S3_NAME:process.env.AWS_S3_NAME,
    AWS_ACCESS_KEY:process.env.AWS_ACCESS_KEY,
    AWS_SecretAccessKey:process.env.AWS_SecretAccessKey,
    get AWS_S3_PREFIX_DOMAIN_NAME(){
        return `https://${process.env.AWS_S3_NAME}.s3.amazonaws.com/`
    },

    /********/
    getRoute: (r: string) => {
        return config.API_PREFIX + r
    },
    route: {
        user: "/user",
        template: "/template",
        app: "/app",
        fileManager: "/fileManager"
    }
}
export default config;
