// ts-glitter/src/config.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.config = exports.saasConfig = exports.ConfigInitial = void 0;
const path_1 = __importDefault(require("path"));
const dotenv_1 = __importDefault(require("dotenv"));
class ConfigInitial {
}
exports.ConfigInitial = ConfigInitial;
ConfigInitial.envPath = path_1.default.resolve(__dirname, `../environments/${process.env.NODE_ENV || 'local'}.env`);
dotenv_1.default.config({
    path: path_1.default.resolve(__dirname, `../environments/${process.env.NODE_ENV || 'local'}.env`),
});
exports.saasConfig = {
    SAAS_NAME: process.env.GLITTER_DB,
    DEF_DEADLINE: 7
};
exports.config = {
    SECRET_KEY: 'dsklkmsadl',
    API_PREFIX: "/api/v1",
    PARAMS_NEED_ENCRYPT_IN_LOG: ['pwd', 'email'],
    PWD_SALT_ROUND: 5,
    SERVER_PORT: '3080',
    LOG_PATH: path_1.default.resolve("../../log"),
    DB_CONN_LIMIT: 1,
    DB_QUEUE_LIMIT: 0,
    DB_URL: process.env.DB_URL,
    DB_PORT: 3306,
    DB_USER: process.env.DB_USER,
    DB_PWD: process.env.DB_PWD,
    DB_NAME: process.env.GLITTER_DB,
    REDIS_URL: process.env.REDIS_URL,
    REDIS_PORT: process.env.REDIS_PORT,
    REDIS_PWD: process.env.REDIS_PWD,
    AWS_S3_NAME: process.env.AWS_S3_NAME,
    AWS_ACCESS_KEY: process.env.AWS_ACCESS_KEY,
    AWS_SecretAccessKey: process.env.AWS_SecretAccessKey,
    get AWS_S3_PREFIX_DOMAIN_NAME() {
        return `https://${process.env.AWS_S3_NAME}.s3.amazonaws.com/`;
    },
    getRoute: (r) => {
        return exports.config.API_PREFIX + r;
    },
    route: {
        user: "/user",
        template: "/template",
        app: "/app",
        fileManager: "/fileManager"
    }
};
exports.default = exports.config;
